go/arrow/flight/flightsql/driver/driver.go

```diff
@@ -37,34 +38,50 @@ import (
 )
 
 type Rows struct {
-	schema        *arrow.Schema
-	records       []arrow.Record
-	currentRecord int
-	currentRow    int
+	schema           *arrow.Schema
+	recordChan       chan arrow.Record
+	currentRecord    arrow.Record
+	currentRecordMux sync.Mutex
+	currentRow       uint64
+	initializedChan  chan bool
+	streamError      error
+}
+
+func newRows() *Rows {
+	return &Rows{
+		recordChan:      make(chan arrow.Record, 1),
+		initializedChan: make(chan bool),
+	}
 }
 
 // Columns returns the names of the columns.
 func (r *Rows) Columns() []string {
-	if len(r.records) == 0 {
+	if r.schema == nil {
 		return nil
 	}
 
 	// All records have the same columns
-	var cols []string
-	for _, c := range r.schema.Fields() {
-		cols = append(cols, c.Name)
+	cols := make([]string, len(r.schema.Fields()))
+	for i, c := range r.schema.Fields() {
+		cols[i] = c.Name
 	}
 
 	return cols
 }
 
+func (r *Rows) releaseRecord() {
+	r.currentRecordMux.Lock()
+	defer r.currentRecordMux.Unlock()
+
+	if r.currentRecord != nil {
+		r.currentRecord.Release()
+		r.currentRecord = nil
+	}
+}
+
 // Close closes the rows iterator.
 func (r *Rows) Close() error {
```

Comment:
what happens if `Close` is called before all the records have been retrieved? How do we cancel the record reader?

Reply:
with these adjustments, Close() now properly cancel() the streaming.